### PR TITLE
Do not hold the guard when creating the iterator.

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe and @geirst PR

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/matching/attribute_limiter.cpp [hunk 1]
```diff
@@ -76,19 +76,19 @@ toString(AttributeLimiter::DiversityCutoffStrategy strategy)
     return (strategy == AttributeLimiter::DiversityCutoffStrategy::STRICT) ? STRICT_STR : LOOSE_STR;
 }
 
-std::unique_ptr<SearchIterator>
-AttributeLimiter::create_search(size_t want_hits, size_t max_group_size, double hit_rate, bool strictSearch)
-{
+search::fef::MatchData &
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/attribute_limiter.cpp:79]: consider returning a pair of blueprint ref and match data ref to not have to touch the internal blueprint without the lock afterwards.

**Review by havardpe — DISMISSED** [review]

File: searchcore/src/vespa/searchcore/proton/matching/attribute_limiter.cpp [hunk 2]
```diff
@@ -105,13 +105,13 @@ AttributeLimiter::create_match_data(size_t want_hits, size_t max_group_size, dou
...
     _match_datas.push_back(layout.createMatchData());
-    return *_match_datas.back();
+    return {*_blueprint, *_match_datas.back()};
 }
 
 std::unique_ptr<SearchIterator>
 AttributeLimiter::create_search(size_t want_hits, size_t max_group_size, double hit_rate, bool strictSearch) {
-    search::fef::MatchData & matchData = create_match_data(want_hits, max_group_size, hit_rate, strictSearch);
-    return _blueprint->createSearch(matchData, strictSearch);
+    auto bp_and_md = create_match_data(want_hits, max_group_size, hit_rate, strictSearch);
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/attribute_limiter.cpp:113]: can do structured capture here: 'auto &[bp, md] = ...'

File: searchcore/src/vespa/searchcore/proton/matching/attribute_limiter.cpp (hunk 1 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/attribute_limiter.cpp]: Fixed

File: searchcore/src/vespa/searchcore/proton/matching/attribute_limiter.cpp (hunk 2 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/attribute_limiter.cpp]: Fixed

**Pull request merged**